src/core/lib/iomgr/tcp_posix.cc

```diff
@@ -671,16 +861,51 @@ ssize_t tcp_send(int fd, const struct msghdr* msg) {
  */
 static bool tcp_write_with_timestamps(grpc_tcp* tcp, struct msghdr* msg,
                                       size_t sending_length,
-                                      ssize_t* sent_length);
+                                      ssize_t* sent_length,
+                                      int additional_flags = 0);
 
 /** The callback function to be invoked when we get an error on the socket. */
 static void tcp_handle_error(void* arg /* grpc_tcp */, grpc_error* error);
 
+static TcpZcpSendRecord* tcp_get_send_zcp_record(grpc_tcp* tcp,
+                                                 grpc_slice_buffer* buf);
+
 #ifdef GRPC_LINUX_ERRQUEUE
+static void process_errors(grpc_tcp* tcp);
+
+static TcpZcpSendRecord* tcp_get_send_zcp_record(grpc_tcp* tcp,
+                                                 grpc_slice_buffer* buf) {
+  TcpZcpSendRecord* zcp_send_record = nullptr;
+  bool use_zerocopy = tcp->tcp_zcp_send_ctx.IsEnabled() &&
+                      tcp->tcp_zcp_send_ctx.ThresholdBytes() < buf->length;
+  if (use_zerocopy) {
+    zcp_send_record = tcp->tcp_zcp_send_ctx.GetSendRecord();
+    if (!zcp_send_record) {
+      process_errors(tcp);
+      zcp_send_record = tcp->tcp_zcp_send_ctx.GetSendRecord();
+    }
+    if (zcp_send_record) {
+      zcp_send_record->PrepareForSends(buf);
+      GPR_DEBUG_ASSERT(buf->count == 0);
+      GPR_DEBUG_ASSERT(buf->length == 0);
+      tcp->outgoing_byte_idx = 0;
+      tcp->outgoing_buffer = nullptr;
+    }
+  }
+  return zcp_send_record;
+}
+
+static void ZerocopyDisableAndWaitForRemaining(grpc_tcp* tcp) {
+  tcp->tcp_zcp_send_ctx.Shutdown();
```

Comment:
A release store at the beginning of a function seems a little weird. Is there any guarantee that the following while loop won't be reordered before this store?

Reply:
The while loop itself performs mutex ops, which presumably would set barriers preventing that from happening.